File: pkg/collection/lib/priority_queue.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library dart.pkg.collection.priority_queue;

import "dart:collection" show SplayTreeSet;

/**
 * A priority queue is a priority based work-list of elements.
 *
 * The queue allows adding elements, and removing them again in priority order.
 */
abstract class PriorityQueue<E> {
  /**
   * Number of elements in the queue.
   */
  int get length;

  /**
   * Whether the queue is empty.
   */
  bool get isEmpty;

  /**
   * Whether the queue has any elements.
   */
  bool get isNotEmpty;

  /**
   * Checks if [object] is in the queue.
   *
   * Returns true if the element is found.
   */
  bool contains(E object);

  /**
   * Adds element to the queue.
   *
   * The element will become the next to be removed by [removeFirst]
   * when all elements with higher priority have been removed.
   */
  void add(E element);

  /**
   * Adds all [elements] to the queue.
   */
  void addAll(Iterable<E> elements);

  /**
   * Returns the next element that will be returned by [removeFirst].
   *
   * The element is not removed from the queue.
   *
   * The queue must not be empty when this method is called.
   */
  E get first;

  /**
   * Removes and returns the element with the highest priority.
   *
   * Repeatedly calling this method, without adding element in between,
   * is guaranteed to return elements in non-decreasing order as, specified by
   * [comparison].
   *
   * The queue must not be empty when this method is called.
   */
  E removeFirst();

  /**
   * Removes an element that compares equal to [element] in the queue.
   *
   * Returns true if an element is found and removed,
   * and false if no equal element is found.
   */
  bool remove(E element);

  /**
   * Removes all the elements from this queue and returns them.
   *
   * The returned iterable has no specified order.
   */
  Iterable<E> removeAll();

  /**
   * Removes all the elements from this queue.
   */
  void clear();

  /**
   * Returns a list of the elements of this queue in priority order.
   *
   * The queue is not modified.
   *
   * The order is the order that the elements would be in if they were
   * removed from this queue using [removeFirst].
   */
  List<E> toList();

  /**
   * Return a comparator based set using the comparator of this queue.
   *
   * The queue is not modified.
   *
   * The returned [Set] is currently a [SplayTreeSet],
   * but this may change as other ordered sets are implemented.
   *
   * The set contains all the elements of this queue.
   * If an element occurs more than once in the queue,
   * the set will contain it only once.
   */
  Set<E> toSet();
}

/**
 * Heap based priority queue.
 *
 * The elements are kept in a heap structure,
 * where the element with the highest priority is immediately accessible,
 * and modifying a single element takes
 * logarithmic time in the number of elements on average.
 *
 * * The [add] and [removeFirst] operations take amortized logarithmic time,
 *   O(log(n)), but may occasionally take linear time when growing the capacity
 *   of the heap.
 * * The [addAll] operation works as doing repeated [add] operations.
 * * The [first] getter takes constant time, O(1).
 * * The [clear] and [removeAll] methods also take constant time, O(1).
 * * The [contains] and [remove] operations may need to search the entire
 *   queue for the elements, taking O(n) time.
 * * The [toList] operation effectively sorts the elements, taking O(n*log(n))
 *   time.
 * * The [toSet] operation effectively adds each element to the new set, taking
 *   an expected O(n*log(n)) time.
 */
class HeapPriorityQueue<E> implements PriorityQueue<E> {
  /**
   * Initial capacity of a queue when created, or when added to after a [clear].
   *
   * Number can be any positive value. Picking a size that gives a whole
   * number of "tree levels" in the heap is only done for aesthetic reasons.
   */
  static const int _INITIAL_CAPACITY = 7;

  /**
   * The comparison being used to compare the priority of elements.
   */
  final Comparator comparison;

  /**
   * List implementation of a heap.
   */
  List<E> _queue = new List<E>(_INITIAL_CAPACITY);

  /**
   * Number of elements in queue.
   *
   * The heap is implemented in the first [_length] entries of [_queue].
   */
  int _length = 0;

  /**
   * Create a new priority queue.
   *
   * The [comparison] is a [Comparator] used to compare the priority of
   * elements. An element that compares as less than another element has
   * a higher priority.
   *
   * If [comparison] is omitted, it defaults to [Comparable.compare].
   */
  HeapPriorityQueue([int comparison(E e1, E e2)])
      : comparison = (comparison != null) ? comparison : Comparable.compare;

  void add(E element) {
    _add(element);
  }

  void addAll(Iterable<E> elements) {
    for (E element in elements) {
      _add(element);
    }
  }

  void clear() {
    _queue = const [];
    _length = 0;
  }

  bool contains(E object) {
    return _locate(object) >= 0;
  }

  E get first {
    if (_length == 0) throw new StateError("No such element");
    return _queue[0];
  }

  bool get isEmpty => _length == 0;

  bool get isNotEmpty => _length != 0;

  int get length => _length;

  bool remove(E element) {
    int index = _locate(element);
    if (index < 0) return false;
    E last = _removeLast();
    if (index < _length) {
      int comp = comparison(last, element);
      if (comp <= 0) {
        _bubbleUp(last, index);
      } else {
        _bubbleDown(last, index);
      }
    }
    return true;
  }

  Iterable<E> removeAll() {
    List<E> result = _queue;
    int length = _length;
    _queue = const [];
    _length = 0;
    return result.take(length);
  }

  E removeFirst() {
    if (_length == 0) throw new StateError("No such element");
    E result = _queue[0];
    E last = _removeLast();
    if (_length > 0) {
      _bubbleDown(last, 0);
    }
    return result;
  }

  List<E> toList() {
    List<E> list = new List<E>()..length = _length;
    list.setRange(0, _length, _queue);
    list.sort(comparison);
    return list;
  }

  Set<E> toSet() {
    Set<E> set = new SplayTreeSet<E>(comparison);
    for (int i = 0; i < _length; i++) {
      set.add(_queue[i]);
    }
    return set;
  }

  /**
   * Returns some representation of the queue.
   *
   * The format isn't significant, and may change in the future.
   */
  String toString() {
    return _queue.take(_length).toString();
  }

  /**
   * Add element to the queue.
   *
   * Grows the capacity if the backing list is full.
   */
  void _add(E element) {
    if (_length == _queue.length) _grow();
    _bubbleUp(element, _length++);
  }

  /**
   * Find the index of an object in the heap.
   *
   * Returns -1 if the object is not found.
   */
  int _locate(E object) {
    if (_length == 0) return -1;
    // Count positions from one instad of zero. This gives the numbers
    // some nice properties. For example, all right children are odd,
    // their left sibling is even, and the parent is found by shifting
    // right by one.
    // Valid range for position is [1.._length], inclusive.
    int position = 1;
    // Pre-order depth first search, omit child nodes if the current
    // node has lower priority than [object], because all nodes lower
    // in the heap will also have lower priority.
    do {
      int index = position - 1;
      E element = _queue[index];
      int comp = comparison(element, object);
      if (comp == 0) return index;
      if (comp < 0) {
        // Element may be in subtree.
        // Continue with the left child, if it is there.
        int leftChildPosition = position * 2;
        if (leftChildPosition <= _length) {
          position = leftChildPosition;
          continue;
        }
      }
      // Find the next right sibling or right ancestor sibling.
      do {
        while (position.isOdd) {
          // While position is a right child, go to the parent.
          position >>= 1;
        }
        // Then go to the right sibling of the left-child.
        position += 1;
      } while (position > _length);  // Happens if last element is a left child.
    } while (position != 1);  // At root again. Happens for right-most element.
    return -1;
  }

  E _removeLast() {
    int newLength = _length - 1;
    E last = _queue[newLength];
    _queue[newLength] = null;
    _length = newLength;
    return last;
  }

  /**
   * Place [element] in heap at [index] or above.
   *
   * Put element into the empty cell at `index`.
   * While the `element` has higher priority than the
   * parent, swap it with the parent.
   */
  void _bubbleUp(E element, int index) {
    while (index > 0) {
      int parentIndex = (index - 1) ~/ 2;
      E parent = _queue[parentIndex];
      if (comparison(element, parent) > 0) break;
      _queue[index] = parent;
      index = parentIndex;
    }
    _queue[index] = element;
  }

  /**
   * Place [element] in heap at [index] or above.
   *
   * Put element into the empty cell at `index`.
   * While the `element` has lower priority than either child,
   * swap it with the highest priority child.
   */
  void _bubbleDown(E element, int index) {
    int rightChildIndex = index * 2 + 2;
    while (rightChildIndex < _length) {
      int leftChildIndex = rightChildIndex - 1;
      E leftChild = _queue[leftChildIndex];
      E rightChild = _queue[rightChildIndex];
      int comp = comparison(leftChild, rightChild);
      int minChildIndex;
      E minChild;
      if (comp < 0) {
        minChild = leftChild;
        minChildIndex = leftChildIndex;
      } else {
        minChild = rightChild;
        minChildIndex = rightChildIndex;
      }
      comp = comparison(element, minChild);
      if (comp <= 0) {
        _queue[index] = element;
        return;
      }
      _queue[index] = minChild;
      index = minChildIndex;
      rightChildIndex = index * 2 + 2;
    }
    int leftChildIndex = rightChildIndex - 1;
    if (leftChildIndex < _length) {
      E child = _queue[leftChildIndex];
      int comp = comparison(element, child);
      if (comp > 0) {
        _queue[index] = child;
        index = leftChildIndex;
      }
    }
    _queue[index] = element;
  }

  /**
   * Grows the capacity of the list holding the heap.
   *
   * Called when the list is full.
   */
  void _grow() {
    int newCapacity = _queue.length * 2 + 1;
    if (newCapacity < _INITIAL_CAPACITY) newCapacity = _INITIAL_CAPACITY;
    List<E> newQueue = new List<E>(newCapacity);
    newQueue.setRange(0, _length, _queue);
    _queue = newQueue;
  }
}

File: pkg/dev_compiler/test/codegen/expect/collection/src/priority_queue.js
dart_library.library('collection/src/priority_queue', null, /* Imports */[
  'dart/_runtime',
  'dart/core',
  'dart/collection'
], /* Lazy imports */[
], function(exports, dart, core, collection) {
  'use strict';
  let dartx = dart.dartx;
  const PriorityQueue$ = dart.generic(function(E) {
    class PriorityQueue extends core.Object {
      static new(comparison) {
        return new (HeapPriorityQueue$(E))(comparison);
      }
    }
    dart.setSignature(PriorityQueue, {
      constructors: () => ({new: [PriorityQueue$(E), [], [dart.functionType(core.int, [E, E])]]})
    });
    return PriorityQueue;
  });
  let PriorityQueue = PriorityQueue$();
  const _queue = Symbol('_queue');
  const _length = Symbol('_length');
  const _add = Symbol('_add');
  const _locate = Symbol('_locate');
  const _removeLast = Symbol('_removeLast');
  const _bubbleUp = Symbol('_bubbleUp');
  const _bubbleDown = Symbol('_bubbleDown');
  const _grow = Symbol('_grow');
  const HeapPriorityQueue$ = dart.generic(function(E) {
    class HeapPriorityQueue extends core.Object {
      HeapPriorityQueue(comparison) {
        if (comparison === void 0) comparison = null;
        this[_queue] = core.List$(E).new(HeapPriorityQueue$()._INITIAL_CAPACITY);
        this.comparison = dart.as(comparison != null ? comparison : core.Comparable.compare, core.Comparator);
        this[_length] = 0;
      }
      add(element) {
        dart.as(element, E);
        this[_add](element);
      }
      addAll(elements) {
        dart.as(elements, core.Iterable$(E));
        for (let element of elements) {
          this[_add](element);
        }
      }
      clear() {
        this[_queue] = dart.const(dart.list([], E));
        this[_length] = 0;
      }
      contains(object) {
        dart.as(object, E);
        return dart.notNull(this[_locate](object)) >= 0;
      }
      get first() {
        if (this[_length] == 0) dart.throw(new core.StateError("No such element"));
        return this[_queue][dartx.get](0);
      }
      get isEmpty() {
        return this[_length] == 0;
      }
      get isNotEmpty() {
        return this[_length] != 0;
      }
      get length() {
        return this[_length];
      }
      remove(element) {
        dart.as(element, E);
        let index = this[_locate](element);
        if (dart.notNull(index) < 0) return false;
        let last = this[_removeLast]();
        if (dart.notNull(index) < dart.notNull(this[_length])) {
          let comp = dart.dcall(this.comparison, last, element);
          if (dart.notNull(comp) <= 0) {
            this[_bubbleUp](last, index);
          } else {
            this[_bubbleDown](last, index);
          }
        }
        return true;
      }
      removeAll() {
        let result = this[_queue];
        let length = this[_length];
        this[_queue] = dart.const(dart.list([], E));
        this[_length] = 0;
        return result[dartx.take](length);
      }
      removeFirst() {
        if (this[_length] == 0) dart.throw(new core.StateError("No such element"));
        let result = this[_queue][dartx.get](0);
        let last = this[_removeLast]();
        if (dart.notNull(this[_length]) > 0) {
          this[_bubbleDown](last, 0);
        }
        return result;
      }
      toList() {
        let list = core.List$(E).new();
        list[dartx.length] = this[_length];
        list[dartx.setRange](0, this[_length], this[_queue]);
        list[dartx.sort](dart.as(this.comparison, __CastType0));
        return list;
      }
      toSet() {
        let set = new (collection.SplayTreeSet$(E))(dart.as(this.comparison, dart.functionType(core.int, [E, E])));
        for (let i = 0; i < dart.notNull(this[_length]); i++) {
          set.add(this[_queue][dartx.get](i));
        }
        return set;
      }
      toString() {
        return dart.toString(this[_queue][dartx.take](this[_length]));
      }
      [_add](element) {
        dart.as(element, E);
        if (this[_length] == this[_queue][dartx.length]) this[_grow]();
        this[_bubbleUp](element, (() => {
          let x = this[_length];
          this[_length] = dart.notNull(x) + 1;
          return x;
        })());
      }
      [_locate](object) {
        dart.as(object, E);
        if (this[_length] == 0) return -1;
        let position = 1;
        do {
          let index = position - 1;
          let element = this[_queue][dartx.get](index);
          let comp = dart.dcall(this.comparison, element, object);
          if (comp == 0) return index;
          if (dart.notNull(comp) < 0) {
            let leftChildPosition = position * 2;
            if (leftChildPosition <= dart.notNull(this[_length])) {
              position = leftChildPosition;
              continue;
            }
          }
          do {
            while (dart.notNull(position[dartx.isOdd])) {
              position = position >> 1;
            }
            position = position + 1;
          } while (position > dart.notNull(this[_length]));
        } while (position != 1);
        return -1;
      }
      [_removeLast]() {
        let newLength = dart.notNull(this[_length]) - 1;
        let last = this[_queue][dartx.get](newLength);
        this[_queue][dartx.set](newLength, null);
        this[_length] = newLength;
        return last;
      }
      [_bubbleUp](element, index) {
        dart.as(element, E);
        while (dart.notNull(index) > 0) {
          let parentIndex = ((dart.notNull(index) - 1) / 2)[dartx.truncate]();
          let parent = this[_queue][dartx.get](parentIndex);
          if (dart.notNull(dart.dcall(this.comparison, element, parent)) > 0) break;
          this[_queue][dartx.set](index, parent);
          index = parentIndex;
        }
        this[_queue][dartx.set](index, element);
      }
      [_bubbleDown](element, index) {
        dart.as(element, E);
        let rightChildIndex = dart.notNull(index) * 2 + 2;
        while (rightChildIndex < dart.notNull(this[_length])) {
          let leftChildIndex = rightChildIndex - 1;
          let leftChild = this[_queue][dartx.get](leftChildIndex);
          let rightChild = this[_queue][dartx.get](rightChildIndex);
          let comp = dart.dcall(this.comparison, leftChild, rightChild);
          let minChildIndex = null;
          let minChild = null;
          if (dart.notNull(comp) < 0) {
            minChild = leftChild;
            minChildIndex = leftChildIndex;
          } else {
            minChild = rightChild;
            minChildIndex = rightChildIndex;
          }
          comp = dart.dcall(this.comparison, element, minChild);
          if (dart.notNull(comp) <= 0) {
            this[_queue][dartx.set](index, element);
            return;
          }
          this[_queue][dartx.set](index, minChild);
          index = minChildIndex;
          rightChildIndex = dart.notNull(index) * 2 + 2;
        }
        let leftChildIndex = rightChildIndex - 1;
        if (leftChildIndex < dart.notNull(this[_length])) {
          let child = this[_queue][dartx.get](leftChildIndex);
          let comp = dart.dcall(this.comparison, element, child);
          if (dart.notNull(comp) > 0) {
            this[_queue][dartx.set](index, child);
            index = leftChildIndex;
          }
        }
        this[_queue][dartx.set](index, element);
      }
      [_grow]() {
        let newCapacity = dart.notNull(this[_queue][dartx.length]) * 2 + 1;
        if (dart.notNull(newCapacity) < dart.notNull(HeapPriorityQueue$()._INITIAL_CAPACITY)) newCapacity = HeapPriorityQueue$()._INITIAL_CAPACITY;
        let newQueue = core.List$(E).new(newCapacity);
        newQueue[dartx.setRange](0, this[_length], this[_queue]);
        this[_queue] = newQueue;
      }
    }
    HeapPriorityQueue[dart.implements] = () => [PriorityQueue$(E)];
    dart.setSignature(HeapPriorityQueue, {
      constructors: () => ({HeapPriorityQueue: [HeapPriorityQueue$(E), [], [dart.functionType(core.int, [E, E])]]}),
      methods: () => ({
        add: [dart.void, [E]],
        addAll: [dart.void, [core.Iterable$(E)]],
        clear: [dart.void, []],
        contains: [core.bool, [E]],
        remove: [core.bool, [E]],
        removeAll: [core.Iterable$(E), []],
        removeFirst: [E, []],
        toList: [core.List$(E), []],
        toSet: [core.Set$(E), []],
        [_add]: [dart.void, [E]],
        [_locate]: [core.int, [E]],
        [_removeLast]: [E, []],
        [_bubbleUp]: [dart.void, [E, core.int]],
        [_bubbleDown]: [dart.void, [E, core.int]],
        [_grow]: [dart.void, []]
      })
    });
    HeapPriorityQueue._INITIAL_CAPACITY = 7;
    return HeapPriorityQueue;
  });
  let HeapPriorityQueue = HeapPriorityQueue$();
  const __CastType0$ = dart.generic(function(E) {
    const __CastType0 = dart.typedef('__CastType0', () => dart.functionType(core.int, [E, E]));
    return __CastType0;
  });
  let __CastType0 = __CastType0$();
  // Exports:
  exports.PriorityQueue$ = PriorityQueue$;
  exports.PriorityQueue = PriorityQueue;
  exports.HeapPriorityQueue$ = HeapPriorityQueue$;
  exports.HeapPriorityQueue = HeapPriorityQueue;
});
